Fuse searchbox

alexandercyu opened the pull request:
Added code to make search input field functional with STAC queryables values

When typing search terms the results will be added to a hidden list below the search input field that expands when there are search results.  Clicking a result will copy list value into search input field.

alexandercyu committed:
Update build.py

Add defaults for stac catalog url

alexandercyu committed:
Update stac-ui.css

Add styles for search input box and search result display list

alexandercyu committed:
Update searchbox.html

Change search result to display in a list instead of a div

alexandercyu committed:
Create catalog-search.js

- Copy code over from Misha's fuse search demo
- Add code to build list elements to display result
- Add function to clear list every time a new search term is entered
- Add code to swap aria-expanded status so it can be used in CSS for styles

alexandercyu committed:
Update sample.js

Add function call to clear li list items  from list ul

alexandercyu committed:
Update sample.html

Add fuse search CDN link

alexandercyu committed:
Update catalog-search.js

Fixed duplicate id problem

mishaschwartz commented on build.py:
```diff
@@ -42,16 +42,23 @@ def build(build_directory, clean=True):
     parser.add_argument(
         "-b",
-        "--build-directory",
+        "--build_directory",
```
Leave as hyphens. 

Hyphens are expected for CLI options

alexandercyu replied:
Hyphens replaced

mishaschwartz commented on build.py:
```diff
@@ -42,16 +42,23 @@ def build(build_directory, clean=True):
 
+    parser.add_argument(
+        "-u",
+        "--stac_catalog_url",
```
hyphens

alexandercyu replied:
Hyphens replaced

mishaschwartz commented on build.py:
```diff
@@ -42,16 +42,23 @@ def build(build_directory, clean=True):
+        "-u",
+        "--stac_catalog_url",
+        default="https://redoak.cs.toronto.edu/stac", #Keep for now. URL should not be U of T specific
+        help="location on disk to write built templates to.",
```
change

alexandercyu replied:
Changed help to be appropriate to the entry.

mishaschwartz commented on templates/site/js/catalog-search.js:
```diff
@@ -0,0 +1,73 @@
+    const resp = await fetch("{{ stac_catalog_url }}/collections")
+    const json = await resp.json()
+    const collection_ids = json.collections.map(collection => collection.id)
+    return await Promise.all(collection_ids.map(async (id) => {
```
You should just query the `/queryables` endpoint for the whole catalog unless specific collections are being searched in which case this function should take those collection ids as inputs.

alexandercyu replied:
Changed to query the /queryables endpoint for the whole catalog.

Also attempted to change the function to take collection IDs as inputs from the default dropdown, but the dropdown needs to be built with the collection IDs as values first.

mishaschwartz commented on templates/site/js/catalog-search.js:
```diff
@@ -0,0 +1,73 @@
+            distance: 10,
+            includeMatches: true
+        });
+        console.log("search is ready") // TODO: replace this with some proper message to the user
```
Search should be disabled until this is ready

alexandercyu replied:
Replaced with function that will removed the 'disabled' attribute placed on the search input box by default.  Typing in the search input box is disabled until the fuse search is finished preparing.

mishaschwartz reviewed (changes requested):
Text in suggestions does not match the design:
- text is too small
- font is different from font in the searchbox
- padding above and below is not balanced

Expanding the suggestion box moves everything below it down, selecting a word does not move everything back:

Initial:
![image](https://github.com/user-attachments/assets/9c0016f6-5927-4170-9942-afe9e242c7c6)

Expanded:
![image](https://github.com/user-attachments/assets/5d1eaf30-02f9-47ee-b61d-a8cafd3586ac)

After selection:
![image](https://github.com/user-attachments/assets/ad260015-c17e-40c9-8eb3-a10748bf69da)

Selecting a value from the suggestion box does not copy it into the text box (see the "After selection" picture above).

Suggestion box does not fully disappear when you type something and then delete what you've typed:

![image](https://github.com/user-attachments/assets/4d916fba-736f-4947-ad61-d6b831be66f5)

Suggestions are duplicated:

![image](https://github.com/user-attachments/assets/6a8a7d20-580a-436a-a623-e04f02761d1f)

alexandercyu commented:
> Text in suggestions does not match the design:
> 
>     * text is too small
> 
>     * font is different from font in the searchbox
> 
>     * padding above and below is not balanced
> 
> 
> Expanding the suggestion box moves everything below it down, selecting a word does not move everything back:
> 
> Initial: ![image](https://private-user-images.githubusercontent.com/4380924/418647247-9c0016f6-5927-4170-9942-afe9e242c7c6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEwMjgxMzEsIm5iZiI6MTc0MTAyNzgzMSwicGF0aCI6Ii80MzgwOTI0LzQxODY0NzI0Ny05YzAwMTZmNi01OTI3LTQxNzAtOTk0Mi1hZmU5ZTI0MmM3YzYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDNUMTg1MDMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzJiNGE5MDU4NTMxNmEzYzExZjM2MDcyN2NjYmNiYzM0YzUzMzgxYzBkYTdiMWVkYjI4MTQzNzBkZDNjZDZlZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.UC7tPP8URUAj7O_qdiauBbETtkGChML2h24rPMWsoZI)
> 
> Expanded: ![image](https://private-user-images.githubusercontent.com/4380924/418647498-5d1eaf30-02f9-47ee-b61d-a8cafd3586ac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEwMjgxMzEsIm5iZiI6MTc0MTAyNzgzMSwicGF0aCI6Ii80MzgwOTI0LzQxODY0NzQ5OC01ZDFlYWYzMC0wMmY5LTQ3ZWUtYjYxZC1hOGNhZmQzNTg2YWMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDNUMTg1MDMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTg3NDk1YmI5MWUwNjRkMmRiZmRkNWY5NmUwZDc3N2VlNzZiYzY0ZmJjZDA4MWZmMjkyZDdlMzVmZjBlZDkwOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.7HO3nII5UVMqM0jsys3MFpE7evg1dUXqdxyy0Xsnu_c)
> 
> After selection: ![image](https://private-user-images.githubusercontent.com/4380924/418647647-ad260015-c17e-40c9-8eb3-a10748bf69da.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEwMjgxMzEsIm5iZiI6MTc0MTAyNzgzMSwicGF0aCI6Ii80MzgwOTI0LzQxODY0NzY0Ny1hZDI2MDAxNS1jMTdlLTQwYzktOGViMy1hMTA3NDhiZjY5ZGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDNUMTg1MDMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzU1ZDlkZWEzZGMyNDhhYzU5Mzk4ODZkYjBhYjUzYWJmNDM3ZmM4MjNmMTQxNjc5YzNjMzkxNjZmZmM0NGZiMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.qQJl2DhZEvPxJqsyfXfeDB6Wl6657-riQlFjLtVLhKQ)
> 
> Selecting a value from the suggestion box does not copy it into the text box (see the "After selection" picture above).
> 
> Suggestion box does not fully disappear when you type something and then delete what you've typed:
> 
> ![image](https://private-user-images.githubusercontent.com/4380924/418649849-4d916fba-736f-4947-ad61-d6b831be66f5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEwMjgxMzEsIm5iZiI6MTc0MTAyNzgzMSwicGF0aCI6Ii80MzgwOTI0LzQxODY0OTg0OS00ZDkxNmZiYS03MzZmLTQ5NDctYWQ2MS1kNmI4MzFiZTY2ZjUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDNUMTg1MDMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDA5NWEwNDg3ODcxZGExMjMzNTNiOGNjYjdkOTk5NWIxOTRlN2QwODYxYmI2Y2IwMGE4NjQ5MDVkMjdkZGNmOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.f8IeDdrlaLaiA1QdKggyjRJdzSWag8lHErpCZWbbWkU)
> 
> Suggestions are duplicated:
> 
> ![image](https://private-user-images.githubusercontent.com/4380924/418650896-6a8a7d20-580a-436a-a623-e04f02761d1f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEwMjgxMzEsIm5iZiI6MTc0MTAyNzgzMSwicGF0aCI6Ii80MzgwOTI0LzQxODY1MDg5Ni02YThhN2QyMC01ODBhLTQzNmEtYTYyMy1lMDRmMDI3NjFkMWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDNUMTg1MDMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTZmNTdmYWZkMTJjMmM0ZGNjMmFjYWM1ZTY0MzFkN2JkMTM3MjkzYTkzZGY0NDgxMjBiNzlmZjNmNGU3YmI5YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.cmGiCwmLfA40Ix5bDC8_XnMIWlwbhckpRBJrh8onru4)

- Text in search results changed to H5, as per design
- Search results area now allows adjacent elements to move back when search results are collapsed
- Selecting a value from the search results now puts that value in the search box
- Attempted to fix empty search result box showing when erasing entered text in search box

TODO
- hide empty search result box when search term is erased
- Change style of text when search result term is selected and copied into search box to be like the design
- Investigate duplicate results
- Disable search box until fuse search is ready

alexandercyu committed:
Update build.py

- hyphens put back
- help entry changed to be appropriate to parser entry

alexandercyu committed:
Update stac-ui.css

Changed styles so it doesn't result in shifting adjacent elements

alexandercyu committed:
Update checkbox.html

Add value attribute to checkbox

alexandercyu committed:
Update dropdown.html

Add value to checkbox (set as label value for now)

alexandercyu committed:
Update catalog-search.js

- add function to get all collection values from checked checkboxes
- modify getWordList() to query the queryables endpoint for the entire catalog
- keep commented code in getWordList() for now
- add function to collapse search results
- add code to toggle search results expanding and collapsing

alexandercyu commented:
> TODO
>
>    - hide empty search result box when search term is erased
>    - Change style of text when search result term is selected and copied into search box to be like the design
>    - Investigate duplicate results
>    - Disable search box until fuse search is ready

- Returned result text style changed
- Empty result box hidden when search term is erased
- Added disable search box by default and then remove disable when fuse search is ready

alexandercyu committed:
Update stac-ui.css

Added style for returned results in search box

alexandercyu committed:
Update catalog-search.js

- Removed console message for when fuse search is ready
- Added default attribute to disable search box by default
- Added function to remove disable attribute from search box when fuse search is ready
- moved makeFuse() function call to sample.js
- added condition to only expand results panel and get results if there is an input value. Closes results panel otherwise
- added function to remove returned result style - text will not be black and background of search box will not be white unless there is a returned result

alexandercyu committed:
Update sample.js

- Moved makeFuse() function call here
- add function to disable search box initally
- added function to remove styles for returned results that are copied into the search box

mishaschwartz reviewed (changes requested):
- padding above and below is still not balanced (see https://github.com/DACCS-Climate/stac-search-ui/pull/6#pullrequestreview-2654575370)

![image](https://github.com/user-attachments/assets/341d0ac7-578a-4bcd-b5fe-661b0605d346)

- empty suggestion box is awkward size. Either make it disappear entirely or add a placeholder like "No suggestions found"

![image](https://github.com/user-attachments/assets/c903164c-cd43-4e28-939d-cef970223b26)

mishaschwartz commented:
TODO:

- [x] style fixes (see review above)
- [x] handle duplicates in suggestions
- [x] handle selecting subset of collections
- [x] limit number of suggestions that are presented to the user initially

alexandercyu committed:
Update catalog-search.js

- Added condition to only show results panel if there are results from the queryables endpoint
- NOTE: Still testing: Add function to move focus to the first result returned in the results panel when the search box detects a down arrow pressed

alexandercyu committed:
Update stac-ui.css

- Change display type to display: flex
- Add the display flex style to the search-output-container
- Put a container around the result list so additional styles can be used
- style result list like the design

alexandercyu committed:
Update searchbox.html

- put div around result list
- add "autocomplete = off"  to turn off browser default autocomplete suggestion

alexandercyu committed:
Update sample.js

Add function call to focus on first item in result list

alexandercyu committed:
Update sample.js

Change how focusOnResults() function is called

alexandercyu committed:
Update catalog-search.js

- use fuse object to build the search return list instead of the flat list
- add attributes for stac queryable keys and stac queryable object types (enum, etc.)
- reduce distance for fuse search to make it more precise

alexandercyu committed:
Update catalog-search.js

- Add function to recursively find queryable entries that are nested deep in objects of type "anyOf"
- Add the recursively retrieved entries to word list generated for fuse
- Change fuse options so it will have better matches

alexandercyu commented:
> TODO:

>     * [x]  handle selecting subset of collections

Note: Values in subset of collections are added to the queryables list for fuse but need specific keys added to the fuse search "keys" option in order to display them in the results.

alexandercyu committed:
Update catalog-search.js

Add nested results under "anyOf" instead of the key value

alexandercyu committed:
Update catalog-search.js

Added extra keys for fuse to look for

alexandercyu committed:
Update catalog-search.js

Added extra key for fuse to search

alexandercyu committed:
Update catalog-search.js

Build path to string item
NOTE: missing top node key
: adds multiple node keys if they are at the same hierarchy

alexandercyu committed:
Update catalog-search.js

Latest changes

mishaschwartz committed:
recursive function to find queryables for fusejs

alexandercyu committed:
Merge pull request #7 from DACCS-Climate/recursive-queryables

recursive function to find queryables for fusejs

alexandercyu committed:
Update catalog-search.js

Remove commented code and unneeded code

alexandercyu committed:
Merge pull request #8 from DACCS-Climate/find-nested-recursive

Find nested recursive

alexandercyu committed:
Update catalog-search.js

Remove unneeded console logs

mishaschwartz commented on templates/site/js/catalog-search.js:
```diff
@@ -0,0 +1,164 @@
+    fuseDictionary().then(queryables => {
+        fuse = new Fuse(queryables, {
+            keys: ["key"],
+            threshold: 0.2, // TODO: experiment with threshold and distance values to get best results
```
are we done experimenting or do we need to keep this comment?

alexandercyu replied:
I think we're done.  The issue before was fixed with changing the distance.

Removed the TODO.

mishaschwartz commented on templates/site/js/catalog-search.js:
```diff
@@ -0,0 +1,164 @@
+        fuse = new Fuse(queryables, {
+            keys: ["key"],
+            threshold: 0.2, // TODO: experiment with threshold and distance values to get best results
+            distance: 100,
```
this is the default, we don't need to specify this.

alexandercyu replied:
Removed the default option.

mishaschwartz commented on templates/site/js/catalog-search.js:
```diff
@@ -0,0 +1,164 @@
+    const resultList = document.getElementById("suggestedWordOutputList");
+
+    var resultLinkArray = resultList.querySelectorAll('li h5 a')
+    //console.log(resultLinkArray);
```
remove

alexandercyu replied:
Removed the comment.

mishaschwartz reviewed (changes requested):
Clicking on an option in the dropdown adds the wrong value to the search box

alexandercyu committed:
Update catalog-search.js

- Remove comment about threshold
- Remove default distance setting for fuse
- remove unneeded comment for setting focus on result set

mishaschwartz commented on templates/site/js/catalog-search.js:
```diff
@@ -0,0 +1,162 @@
+    var resultLinkArray = resultList.querySelectorAll('li h5 a')
+    //if(keyPressed === "ArrowDown"){
+        var firstLink = document.getElementById(resultLinkArray[0].id)
+        //console.log(firstLink);
```
Do I really have to comment on every single one? I don't want to have to write a comment like this https://github.com/DACCS-Climate/stac-search-ui/pull/6#discussion_r1990066580 every single time you don't remove debugging code or comments from your PRs.

I need you to look through your PRs and double check them before you ask for a review.

I will review code correctness and style. I will not review unfinished code.

alexandercyu replied:
Removed the unfinished function.

mishaschwartz reviewed (changes requested):
This hasn't been resolved: https://github.com/DACCS-Climate/stac-search-ui/pull/6#pullrequestreview-2676018861

alexandercyu committed:
Update catalog-search.js

Fixed error where the incorrect search return value was being copied into the search input box

alexandercyu committed:
Update catalog-search.js

Removed unfinished function

alexandercyu commented:
> Clicking on an option in the dropdown adds the wrong value to the search box

Fixed the issue.
